### PR TITLE
Allow server rendering (resolve #140)

### LINKED ISSUE #140: Address SSR issues
Local storage shouldn't be used directly (could probably just `get(localStorage, 'token')` with lodash). Are there any other issues? Is there an eslint plugin to catch these issues before pushing them?

### PR DESCRIPTION
**Pull request by trevordmiller**
# Changes

- Don't assume client-side globals are available, so everything should work on the server as well
- Add test to server render all "resources" (things exported that you can import from the `egghead-ui package`), so pull requests will fail if any client-side only logic is added

# Result For User

## New test fails if there are any rendering errors when rendering all resources on the server
screen shot 2017-05-08 at 2 39 05 pm

## After fixing client-side only logic, these pass
screen shot 2017-05-08 at 2 37 52 pm

Note that logs still show which is why you see the warning from prop-types (I'll be fixing this in another story). The tests only fail when there is an exit code, which is what happens if any of the resources can't render on the server.

---

### CODE REVIEW

**Comment by trevordmiller** [comment]
@joelhooks @ijones16 @tayiorbeii Does this look ok? As far as I know, the only other issues that would prevent egghead-rails from upgrading to latest egghead-ui are the small breaking changes in `<Button>` (the `size` values have changed, and `pill` is no longer a prop: https://styleguide.egghead.io/components/Button)

**Comment by trevordmiller** [comment]
Hold off reviewing this, I'm going to add a test that renders a server importing the egghead-ui library to ensure these changes actually fix the problems and prevent adding more in the future

**Comment by trevordmiller** [comment]
@joelhooks @ijones16 @tayiorbeii Ok this is ready for review now. I've updated the description with more info. I started going down a rabbit hole (set up a next server and webdriverIO etc. and then realized all that was needed to test server rendering is no errors on `ReactDOMServer.renderToString` 🤦‍♂️ ). This PR should fix the initial server-rendered runtime errors and prevent new ones in the future.

File: .gitignore [hunk 1]
```diff
@@ -1,7 +1,6 @@
 node_modules/
 lib/
 build/
-src/App/markdown
```
Comment (trevordmiller) [inline .gitignore:4]: Removed a long time ago

File: package.json [hunk 2]
```diff
@@ -9,7 +9,7 @@
     "dev:app:prod": "REACT_APP_EGGHEAD_BASE_URL=https://egghead.io NODE_PATH=./src react-scripts start",
     "test": "NODE_PATH=./src react-scripts test",
     "lint": "NODE_PATH=./src eslint src",
-    "build:package": "NODE_PATH=./src babel src/package -d lib",
+    "build:package": "NODE_PATH=./src babel src/package -d lib -i '**/*.test.js'",
```
Comment (trevordmiller) [inline package.json:12]: Don't compile test files to distributed package

File: package.json [hunk 3]
```diff
@@ -50,7 +50,7 @@
   },
   "dependencies": {
     "axios": "^0.15.3",
-    "chart.js": "^2.5.0",
+    "chart.js": "chartjs/Chart.js",
```
Comment (trevordmiller) [inline package.json:53]: Use chart.js master for now until server rendering is released on npm: https://github.com/chartjs/Chart.js/pull/3909 (merged into master, but not yet released on npm)

File: package.json [hunk 4]
```diff
@@ -76,9 +76,9 @@
     "babel-plugin-transform-object-rest-spread": "^6.23.0",
     "babel-preset-latest": "^6.22.0",
     "babel-preset-react": "^6.22.0",
-    "react": "^15.5.4",
-    "react-dom": "^15.5.4",
-    "react-scripts": "^0.8.4"
+    "react": "^15.4.2",
+    "react-dom": "^15.4.2",
+    "react-scripts": "^0.7.0"
```
Comment (trevordmiller) [inline package.json:81]: Keep all react dependencies consistent with peer dependencies until they we upgrade all react related dependencies all together to react 15.5: https://github.com/eggheadio/egghead-ui/issues/108

File: src/App/components/Resource/index.js [hunk 5]
```diff
@@ -52,19 +52,17 @@ const Resource = ({name, resource}) => (
...
+        title: 'Examples',
+        component: (
+          <div className='mt4'>
+            <Examples 
+              createExamples={resource.createExamples}
+              optOut={resource.optOut}
+            />
+          </div>
+        ),
+      },
```
Comment (trevordmiller) [inline src/App/components/Resource/index.js:65]: Resources now require examples

File: src/App/utils/resourcesByType/index.js [hunk 6]
```diff
@@ -886,10 +890,16 @@ export const resourcesByType = {
...
+        createExamples: () => [
+          login(),
+        ],
       },
 
       logout: {
         useCase: `Removes "token" from local storage and refreshes the page to log the user out.`,
+        createExamples: () => [
+          logout(),
+        ],
```
Comment (trevordmiller) [inline src/App/utils/resourcesByType/index.js:902]: This is very meta, when you go to this resource "example" it will actually log you out 😂

**Review by tayiorbeii — APPROVED** [review]
Your comments are really helpful, thanks for doing that!

**Comment by trevordmiller** [comment]
_For future reference:_

So the gist of it is, a bunch of stuff is available globally (like `setTimeout`, `setInterval` etc.) but the global space is different depending on if the code is being run on the server (node) or browser; things like `localStorage` are actually a shorthand for `window.localStorage`, where `window` only exists when running in the browser. Because egghead-rails runs on the server for the first render, and then on the browser for the next renders, `window` isn't defined so it creates a runtime error. This is why I've added the conditionals like `const window = window || false` and `if (!localStorage) return` etc.

The node global is called `global` which has some of the same things as `window`, but only things that are not specific to the DOM. So you can do `setInterval(...)` in both node and browser (since `global.setInterval` and `window.setInterval` both exist), but you can't do dom specific things like `localStorage(...)` because `global.localStorage` doesn't exist, it's only on `window.localStorage`.

Moral of the story is: globals are messy and it's better to use npm modules that work on the server and browser wherever possible, but in our case, I had to add some gross conditionals so the initial render can occur on the server without any runtime errors

**Pull request merged**